### PR TITLE
[SPARK-38630][K8S] K8s app name label should start and end with alphanumeric char

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to fix `getAppNameLabel` by removing the prefix and suffix `-` character from app name label.

### Why are the changes needed?

Currently, `master/branch-3.3` has this regression.
```
io.fabric8.kubernetes.client.KubernetesClientException:
Failure executing: POST at: https://kubernetes.default.svc/api/v1/namespaces/default/pods.
Message: Pod "..." is invalid:
metadata.labels: Invalid value: "...-tpcds-1000g-parquet-":
a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and
must start and end with an alphanumeric character
```

### Does this PR introduce _any_ user-facing change?

This is a regression at Apache Spark 3.3.0.

### How was this patch tested?

Pass the CI with the newly added test case.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this please, @viirya and @HyukjinKwon ?

**Review by viirya — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Oops. Sorry, I'm going to add more test cases for a corner case~

**Comment by viirya** [comment]
Good catch. Will check again after you make it ready to review.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesConfSuite.scala [hunk 1]
```diff
@@ -244,4 +244,10 @@ class KubernetesConfSuite extends SparkFunSuite {
     assert(KubernetesConf.getAppNameLabel("a" * 64) === "a" * 63)
     assert(KubernetesConf.getAppNameLabel("a" * 253) === "a" * 63)
   }
+
+  test("SPARK-38630: K8s label value should start and end with alphanumeric") {
+    assert(KubernetesConf.getAppNameLabel("-hello-") === "hello")
+    assert(KubernetesConf.getAppNameLabel("a" * 62 + "-aaa") === "a" * 62)
+    assert(KubernetesConf.getAppNameLabel("-" + "a" * 63) === "a" * 62)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/KubernetesConfSuite.scala:251]: The above two cases happen after `StringUtils.abbreviate`.

**Comment by dongjoon-hyun** [comment]
The PR is ready back. Thank you again, @viirya .

**Review by viirya — APPROVED** [review]
lgtm

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala [hunk 2]
```diff
@@ -270,7 +271,7 @@ private[spark] object KubernetesConf {
         .replaceAll("-+", "-"),
       "",
       KUBERNETES_DNSNAME_MAX_LENGTH
-    )
+    ).stripPrefix("-").stripSuffix("-")
```
Comment (Yikun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala:274]: Note for my self or others: k8s label allow `dashes (-), underscores (_), dots (.)`, we alread replaced `underscores (_)`, `dots (.)` by `dashes (-)` before, so strip "-" prefix/suffix is enough in here.

Below tests are also passed:
```
assert(KubernetesConf.getAppNameLabel(".hello.") === "hello")
assert(KubernetesConf.getAppNameLabel("_hello_") === "hello")
```

**Review by Yikun — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya and @Yikun .

All tests passed and I verified K8s IT manually.
```
[info] KubernetesSuite:
[info] - Run SparkPi with no resources (8 seconds, 815 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (8 seconds, 330 milliseconds)
[info] - Run SparkPi with a very long application name. (8 seconds, 354 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (8 seconds, 454 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (8 seconds, 355 milliseconds)
[info] - Run SparkPi with an argument. (8 seconds, 342 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (7 seconds, 344 milliseconds)
[info] - All pods have the same service account by default (8 seconds, 527 milliseconds)
[info] - Run extraJVMOptions check on driver (4 seconds, 282 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (8 seconds, 367 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (8 seconds, 739 milliseconds)
[info] - Run SparkPi with env and mount secrets. (16 seconds, 115 milliseconds)
[info] - Run PySpark on simple pi.py example (8 seconds, 351 milliseconds)
[info] - Run PySpark to test a pyfiles example (10 seconds, 449 milliseconds)
[info] - Run PySpark with memory customization (8 seconds, 520 milliseconds)
[info] - Run in client mode. (6 seconds, 251 milliseconds)
[info] - Start pod creation from template (8 seconds, 369 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (8 seconds, 373 milliseconds)
[info] - Test basic decommissioning (39 seconds, 892 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (40 seconds, 266 milliseconds)
[info] *** Test still running after 2 minutes, 24 seconds: suite name: KubernetesSuite, test name: Test decommissioning with dynamic allocation & shuffle cleanups.
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 40 seconds)
[info] - Test decommissioning timeouts (40 seconds, 381 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 6 seconds)
[info] - Run SparkR on simple dataframe.R example (10 seconds, 350 milliseconds)
```

Merged to master/3.3.

**Pull request closed without merging**